compiler/packages/babel-plugin-react-compiler/src/TypeInference/InferTypes.ts

```diff
@@ -483,30 +483,135 @@ class Unifier {
     }
 
     if (type.kind === 'Phi') {
-      const operands = new Set(type.operands.map(i => this.get(i).kind));
-
-      CompilerError.invariant(operands.size > 0, {
+      CompilerError.invariant(type.operands.length > 0, {
         reason: 'there should be at least one operand',
         description: null,
         loc: null,
         suggestions: null,
       });
-      const kind = operands.values().next().value;
 
-      // there's only one unique type and it's not a type var
-      if (operands.size === 1 && kind !== 'Type') {
-        this.unify(v, type.operands[0]);
+      let candidateType: Type | null = null;
+      for (const operand of type.operands) {
+        const resolved = this.get(operand);
+        if (candidateType === null) {
+          candidateType = resolved;
+        } else if (!typeEquals(resolved, candidateType)) {
+          candidateType = null;
+          break;
+        } // else same type, continue
+      }
+
+      if (candidateType !== null) {
+        this.unify(v, candidateType);
         return;
       }
```

Comment:
the previous logic was incorrect, checking only that the operand type kinds matched. we need to check that the types actually match